### PR TITLE
#31363 format template configure in settings

### LINKED ISSUE #31363: Allow to configure FORMAT Template with string in SETTINGS
**Use case**

Sometimes it is too hard or impossible to deploy format output configuration for template format to directory on all nodes in cluster. Sometimes format is very specific and trivial that it is easier to add it to query.

Currently you need to place it to specific directory and pass path to file with template.
https://github.com/ClickHouse/ClickHouse/blob/a16eda68dc54ff0c3be7362dfec02061074f6e04/tests/queries/0_stateless/00938_template_input_format.sh#L46-L51

**Describe the solution you'd like**
Something like this
``` sql
SELECT
    cluster,
    shard_num,
    replica_num
FROM system.clusters
WHERE host_name = fqdn()
SETTINGS format_schema_rows_template = '|| ${cluster:Quoted} | ${shard_num:Quoted} | ${replica_num:Quoted}||', format_schema_rows_between_delimiter = '\n'
```

We add a separate setting and check that only one is set (filename or string).

### PR DESCRIPTION
**Pull request by Blargian**
Hi Clickhouse team, 
I'm opening a PR for this feature. Look forward to your feedback and fixing any issues. :) 

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The user can now specify the template string directly in the query using `format_schema_rows_template` as an alternative to `format_template_row`. Closes https://github.com/ClickHouse/ClickHouse/issues/31363.

### Documentation entry for user-facing changes

- [X] Documentation is written - English and Russian. (То что написал на русском скорее всего есть ошибки, я не носитель русского... нужно помощь исправить пж)

- docs/en/interfaces/formats.md
- docs/ru/interfaces/formats.md
- docs/en/operations/settings/settings-formats.md

Example usage not yet added, I can add that if needed.

### CODE REVIEW

**Comment by Felixoid** [comment]
Dear @Blargian, the `#` in the branch name breaks the CI because we don't URL-encode it. The fix won't be instant, so please, recreate the PR from another branch.

**Comment by Avogar** [comment]
The problem with CI is the branch name, we have a bug in python script with URL encoding (symbol `#` in the branch name wasn't encoded properly). We will fix it, but it will take time. You can recreate a PR with new branch name or wait until it's fixed

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 5f8565307e40a24247bbdd34cb9c9097f762af31 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59088/5f8565307e40a24247bbdd34cb9c9097f762af31/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59088/5f8565307e40a24247bbdd34cb9c9097f762af31/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by Felixoid** [comment]
I have restored the previous commit to not affect the autoapprove lambda. It should block changes into `tests/ci` directory

**Comment by Blargian** [comment]
@Felixoid thanks for the heads up and also for the fix!

**Comment by Felixoid** [comment]
> Do you maybe know why?

Do you know a joke?

> -- We have good and bad news for you
> -- I'd like to start with a good part, doctor
> -- We'll call decease after you

https://github.com/ClickHouse/ClickHouse/pull/59184

**Comment by Blargian** [comment]
> Do you know a joke?
> 
> > -- We have good and bad news for you
> > -- I'd like to start with a good part, doctor
> > -- We'll call decease after you

Рад служить! )

File: src/Core/Settings.h [hunk 1]
```diff
@@ -1080,6 +1080,7 @@ class IColumn;
     M(String, format_schema, "", "Schema identifier (used by schema-based formats)", 0) \
     M(String, format_template_resultset, "", "Path to file which contains format string for result set (for Template format)", 0) \
```
Comment (Avogar) [inline src/Core/Settings.h:1081]: Let's also add similar setting for resultset template, will be useful to have both

Comment (Avogar) [reply src/Core/Settings.h:1081]: Also, please, add new settings to `SettingsChangesHistory.h`

File: src/Processors/Formats/Impl/TemplateBlockOutputFormat.cpp [hunk 2]
```diff
@@ -213,14 +214,27 @@ void registerOutputFormatTemplate(FormatFactory & factory)
...
-        ParsedTemplateFormatString row_format = ParsedTemplateFormatString(
+        ParsedTemplateFormatString row_format;
+        auto idx_by_name = [&](const String & colName)
+        {
+            return sample.getPositionByName(colName);
+        };
+        if (settings.template_settings.row_format.empty())
+        {
+            row_format = ParsedTemplateFormatString();
+            row_format.parse(settings.template_settings.row_format_schema,idx_by_name);
```
Comment (Avogar) [inline src/Processors/Formats/Impl/TemplateBlockOutputFormat.cpp:225]:
```suggestion
            row_format.parse(settings.template_settings.row_format_schema, idx_by_name);
```

File: src/Core/Settings.h [hunk 3]
```diff
@@ -1080,6 +1080,7 @@ class IColumn;
...
     M(String, format_template_row, "", "Path to file which contains format string for rows (for Template format)", 0) \
+    M(String, format_schema_rows_template, "", "Format string for rows (for Template format)", 0) \
```
Comment (Avogar) [inline src/Core/Settings.h:1083]: Let's rename it to `format_template_row_schema` or `format_template_row_format`. We should keep `format_template_` prefix as it indicates that this setting is related only to format Template

File: src/Core/SettingsChangesHistory.h [hunk 4]
```diff
@@ -102,6 +102,8 @@ static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> sett
               {"function_visible_width_behavior", 0, 1, "We changed the default behavior of `visibleWidth` to be more precise"},
               {"max_estimated_execution_time", 0, 0, "Separate max_execution_time and max_estimated_execution_time"},
               {"iceberg_engine_ignore_schema_evolution", false, false, "Allow to ignore schema evolution in Iceberg table engine"},
+              {"format_template_row_format", "none", "", "Template row format string can be set directly in query"},
+              {"format_template_resultset_format", "none", "", "Template result set format string can be set in query"},
```
Comment (Avogar) [inline src/Core/SettingsChangesHistory.h:106]:
```suggestion
              {"format_template_row_format", "", "", "Template row format string can be set directly in query"},
              {"format_template_resultset_format", "", "", "Template result set format string can be set in query"},
```
No need to use "none", just keep empty

**Comment by Blargian** [comment]
@Avogar the last few tests are all failing on issues related to not being able to find, or remove a `.tmp` version of my shell test. Where does the `.tmp` come from? Is there a specific permission level that test files should be given?

**Comment by Avogar** [comment]
The problem is that you used filenames `00937_template_output_format_row.tmp//00937_template_output_format_resultset.tmp`, the same as in test `00937_template_output_format.sh`, but we run tests in parallel, so one test can create files, another remove them. Also in `Stateless tests flaky check` we run the same test in parallel with itself, so the same problem. Let's use unique filenames using env variable `$CLICKHOUSE_TEST_UNIQUE_NAME`

**Comment by Blargian** [comment]
@Avogar is there some bigger issue with CI at the moment? Fast tests fails since I updated my branch from master this morning. It was okay yesterday.

**Comment by Avogar** [comment]
Test 02720_row_policy_column_with_dots was broken in master by accident, already fixed, so should be ok now

**Review by Avogar — APPROVED** [review]

**Comment by Blargian** [comment]
**Remaining tests failing**:

- test_storage_kafka/test.py::test_system_kafka_consumers_rebalance_mv (#56043)
- 02981_vertical_merges_memory_usage (#59648)

**Comment by Avogar** [comment]
Sorry for waiting, I was on vacation. I will resolve conflicts and merge this PR when tests are finished

**Comment by Blargian** [comment]
> Sorry for waiting, I was on vacation. I will resolve conflicts and merge this PR when tests are finished

No worries, thanks! Hope it was a good vacation :+1:

**Pull request merged**